src/sql/workbench/services/connection/browser/connectionManagementService.ts

```diff
@@ -552,20 +555,30 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 		});
 	}
 
-	private handleConnectionError(connection: interfaces.IConnectionProfile, uri: string, options: IConnectionCompletionOptions, callbacks: IConnectionCallbacks, connectionResult: IConnectionResult) {
+	private async handleConnectionError(connection: interfaces.IConnectionProfile, uri: string, options: IConnectionCompletionOptions, callbacks: IConnectionCallbacks, connectionResult: IConnectionResult) {
 		let connectionNotAcceptedError = nls.localize('connectionNotAcceptedError', "Connection Not Accepted");
 		if (options.showFirewallRuleOnError && connectionResult.errorCode) {
-			return this.handleFirewallRuleError(connection, connectionResult).then(success => {
-				if (success) {
-					options.showFirewallRuleOnError = false;
+			let firewallRuleErrorHandled = await this.handleFirewallRuleError(connection, connectionResult);
+			if (firewallRuleErrorHandled) {
+				options.showFirewallRuleOnError = false;
+				return this.connectWithOptions(connection, uri, options, callbacks);
+			}
+			else {
+				let connectionErrorHandled = await this._errorDiagnosticsService.tryHandleConnectionError(connectionResult.errorCode, connectionResult.errorMessage, connection.providerName, Utils.convertToRpcConnectionProfile(connection, false));
```

Comment:
I would not suggest converting the  profile here, just do that in the mainThread handler (no reason to convert these profiles any earlier than need be, and that helps ensure consistency with what connection profile types we work with internally)

Reply:
I'm confused how to do this, as it appears that the code in the extension provider (where it is being registered), must have the same input type as the one in azdata.proposed, and to pass it into the mainThread handler, you would also have to pass in the same type. So that means that both would either have to be IConnectionProfile or ConnectionProfile,

Reply:
I'm moving it into the ErrorService where it should be.

Reply:
It shouldn't be there either IMO (the service shouldn't be dealing with azdata types either). mainThreadErrorDIagnostics is typically where we'd put something like that (just like mainThreadConnectionManagement is doing)

Reply:
Decided to convert to the SqlExtHostTypes implementation since it has what we need in addition to an easy constructor feature (also will not expose azdata in the service)